modules/standard/DistributedDeque.chpl

```diff
@@ -0,0 +1,1000 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+use Collection;
+
+/*
+  A parallel-safe scalable distributed deque. A deque is a double-ended queue that supports
+  insertion and removal from both ends of the queue, effectively supporting both
+  FIFO, LIFO, and a Total ordering, where the order in which you add them will be
+  the exact order you remove them in; for emphasis, a Deque can be used as a Queue,
+  a Stack, and a List respectively. The deque enforces a 'barrier', which is a method
+  of checking whether you may proceed with an operation, that is wait-free under
+  most cases, lock-free in the worst case, which guarantees scalability. In the
+  barrier, we perform freeze checks (if DEQUE_NO_FREEZE is not enabled), and bounds
+  checking (if a capacity is given). Deque operations are separated into two
+  levels: global and local. At a global level, we use simple fetchAdd and fetchSub
+  counters to denote which local deque we apply our operation to, and at a local level
+  we use an unrolled linked list which further has its own (non-atomic) counter.
+*/
+
+/*
+  Frozen states... If we are FREEZE_UNFROZEN, we are mutable. If we are FREEZE_FROZEN,
+  we are immutable. If we are FREEZE_MARKED, we are in the middle of a state change.
+*/
+private const FREEZE_UNFROZEN = 0;
+private const FREEZE_MARKED = 1;
+private const FREEZE_FROZEN = 2;
+
+/*
+  Size of each unroll block for each local deque node.
+*/
+config param DEQUE_BLOCK_SIZE = 8;
+/*
+    Turn off checks for freezing the data structure.
+*/
+config param DEQUE_NO_FREEZE = false;
+
+// For each node we manage an unroll block. This block needs to also support Deque
+// operations, and as such we maintain our own mini headIdx and tailIdx. Since we can
+// add and remove from either direction, we must also maintain the size ourselves...
+pragma "no doc"
+class LocalDequeNode {

```

Comment:
I'm not sure about merging them. I use constant sized tuples, but truth be told I did not have time to benchmark whether or not increasing the size of the tuples benefits performance (it likely does), and converting to C pointers so that I can double the size of each successor node likely would also yield another significant boost for when we have a large workload. 

Time constraints being as they are, I left them as tuples.